File: orderflow_system/desktop/ui/scanner.js
/* ══════════════════════════════════════════════════════════════════
   Scanner view — one ranked table across every streaming instrument.

   the reference platform's Market Analyzer watches a list and sorts by a column; this is
   the same idea over the order-flow columns this program already computes, so
   "which of my instruments is doing something right now?" is one glance.
   ══════════════════════════════════════════════════════════════════ */

const SCAN = { sort: 'score', view: null, lastKey: '', rows: [], timer: null };

const SCAN_COLUMNS = [
    ['symbol', 'Instrument', 'The instrument. Click a row to switch the whole app to it.'],
    ['last', 'Last', 'Last traded price.'],
    ['chg_pct', 'Δ%', 'Change over the scanner window, from the tape itself (not from candles).'],
    ['volume', 'Volume', 'Traded volume in the window.'],
    ['delta', 'Delta', 'Buy volume minus sell volume in the window.'],
    ['delta_pct', 'Δ% of vol', 'Delta as a percentage of volume — who is winning the tape, size-adjusted.'],
    ['prints_per_s', 'Prints/s', 'Tape speed: prints per second.'],
    ['big_trades', 'Big', 'Adaptive big-print count in the window.'],
    ['sweeps', 'Sweeps', 'Multi-level sweep count.'],
    ['stop_runs', 'Stop runs', 'Stop-run detections.'],
    ['liquidations', 'Liqs', 'Liquidation events seen.'],
    ['imbalances', 'Imb', 'Levels currently showing a stacked imbalance.'],
    ['pressure_buy', 'Book bid %', 'Weighted bid liquidity as a percentage of its own recent normal.'],
    ['pressure_sell', 'Book ask %', 'Weighted ask liquidity as a percentage of its own recent normal.'],
    ['absorption_score', 'Absorb', 'Absorption score: aggression that is failing to move price.'],
    ['depth_executions', 'Eaten', 'Prints that took a large share of the size resting at their price.'],
    ['depth_refills', 'Refilled', 'Levels that came back after being eaten (inferred hidden size).'],
    ['ticks_from_vwap', 'VWAP t', 'Distance from the VWAP in ticks; negative means below it.'],
    ['score', 'Score', 'The stated composite: 0.28·|Δ%| + 0.22·|book edge| + 0.18·absorption + 0.17·depth events + 0.15·tape speed, each capped.'],
];

function scanStyles() {
    if (document.getElementById('scanStyleSheet')) return;
    const st = document.createElement('style');
    st.id = 'scanStyleSheet';
    st.textContent = `
        .scan-wrap { overflow-x: auto; }
        table.scan { width: 100%; border-collapse: collapse; font-size: 12px; }
        table.scan th { text-align: right; padding: 6px 8px; color: var(--dim); cursor: pointer;
                        border-bottom: 1px solid var(--line); white-space: nowrap; font-weight: 600; }
        table.scan th:first-child, table.scan td:first-child { text-align: left; }
        table.scan th:hover { color: var(--fg); }
        table.scan th.sorted { color: var(--fg); text-decoration: underline; }
        table.scan td { text-align: right; padding: 5px 8px; border-bottom: 1px solid var(--line);
                        white-space: nowrap; }
        table.scan tr:hover td { background: rgba(255,255,255,0.03); }
        table.scan td.sym { cursor: pointer; font-weight: 600; }
        .scan-pos { color: var(--up, #35d07f); }
        .scan-neg { color: var(--down, #ff5d6c); }
        .scan-bar { display: inline-block; height: 6px; border-radius: 3px; vertical-align: middle;
                    background: rgba(240,196,84,0.75); min-width: 2px; }
        .scan-note { color: var(--dim); font-size: 11px; margin-top: 6px; }
    `;
    document.head.appendChild(st);
}

function scanEnsureView() {
    if (document.getElementById('scanView')) return;
    scanStyles();
    const rail = document.querySelector('.rail');
    if (rail && !document.querySelector('.nav-item[data-view="scanner"]')) {
        const btn = document.createElement('button');
        btn.className = 'nav-item';
        btn.dataset.view = 'scanner';
        btn.innerHTML = '<span class="nav-icon">\u2263</span> Scanner';
        btn.title = 'One ranked row per streaming instrument — the Market Analyzer view.';
        btn.onclick = () => window.showView && window.showView('scanner');
        const before = rail.querySelector('.nav-item[data-view="trackers"]')
            || rail.querySelector('.nav-item[data-view="logs"]');
        rail.insertBefore(btn, before || null);
    }
    const main = document.querySelector('main.views');
    if (!main) return;
    const section = document.createElement('section');
    section.className = 'view';
    section.dataset.view = 'scanner';
    section.id = 'scanView';
    const head = SCAN_COLUMNS.map(([key, label, tip]) =>
        `<th data-sort="${key}" title="${tip}">${label}</th>`).join('');
    section.innerHTML = `
        <div class="view-head">
            <div class="view-title">Scanner</div>
            <div class="view-sub" id="scanSub">one ranked row per instrument — click a column to sort, a row to switch</div>
            <div class="grow"></div>
            <button class="btn small" id="scanRefresh" title="Refresh the table now">Refresh</button>
        </div>
        <div class="card"><div class="card-body">
            <div class="scan-wrap"><table class="scan">
                <thead><tr>${head}</tr></thead>
                <tbody id="scanRows"><tr><td colspan="19" class="dim">waiting for the engine…</td></tr></tbody>
            </table></div>
            <div class="scan-note" id="scanNote"></div>
        </div></div>`;
    main.appendChild(section);
    section.querySelectorAll('th[data-sort]').forEach((th) => {
        th.onclick = () => { SCAN.sort = th.dataset.sort; SCAN.lastKey = ''; scanRefresh(); };
    });
    section.querySelector('#scanRefresh').onclick = () => { SCAN.lastKey = ''; scanRefresh(); };
}

function scanNum(v, digits = 2) {
    if (v === null || v === undefined || v === '') return '–';
    const n = Number(v);
    if (!isFinite(n)) return '–';
    if (Math.abs(n) >= 1e9) return (n / 1e9).toFixed(2) + 'B';
    if (Math.abs(n) >= 1e6) return (n / 1e6).toFixed(2) + 'M';
    if (Math.abs(n) >= 1e3) return (n / 1e3).toFixed(2) + 'K';
    return n.toFixed(digits);
}

function scanRender(table) {
    const body = document.getElementById('scanRows');
    if (!body) return;
    const rows = table.rows || [];
    if (!rows.length) {
        body.innerHTML = '<tr><td colspan="19" class="dim">the engine is not streaming anything yet</td></tr>';
    } else {
        const maxScore = Math.max(1, ...rows.map((r) => r.score || 0));
        body.innerHTML = rows.map((r) => {
            const pos = (v) => (v || 0) > 0 ? 'scan-pos' : ((v || 0) < 0 ? 'scan-neg' : '');
            const width = Math.round(46 * ((r.score || 0) / maxScore));
            return `<tr>
                <td class="sym" data-symbol="${r.symbol}" title="Switch the app to ${r.symbol}">${r.symbol}</td>
                <td>${scanNum(r.last, 4)}</td>
                <td class="${pos(r.chg_pct)}">${r.chg_pct === null ? '–' : (r.chg_pct > 0 ? '+' : '') + scanNum(r.chg_pct)}</td>
                <td>${scanNum(r.volume, 2)}</td>
                <td class="${pos(r.delta)}">${scanNum(r.delta, 2)}</td>
                <td class="${pos(r.delta_pct)}">${r.delta_pct === null ? '–' : scanNum(r.delta_pct, 1)}</td>
                <td>${scanNum(r.prints_per_s, 1)}</td>
                <td>${r.big_trades || 0}</td>
                <td>${r.sweeps || 0}</td>
                <td>${r.stop_runs || 0}</td>
                <td>${r.liquidations || 0}</td>
                <td>${r.imbalances || 0}</td>
                <td>${scanNum(r.pressure_buy, 0)}</td>
                <td>${scanNum(r.pressure_sell, 0)}</td>
                <td>${scanNum(r.absorption_score, 0)}</td>
                <td>${r.depth_executions || 0}</td>
                <td>${r.depth_refills || 0}</td>
                <td>${r.ticks_from_vwap === null || r.ticks_from_vwap === undefined ? '–' : scanNum(r.ticks_from_vwap, 1)}</td>
                <td title="score ${r.score}"><span class="scan-bar" style="width:${width}px"></span> ${scanNum(r.score, 0)}</td>
            </tr>`;
        }).join('');
        body.querySelectorAll('td.sym').forEach((td) => {
            td.onclick = () => {
                const sym = td.dataset.symbol;
                const sel = document.getElementById('symbolSelect');
                if (sel && sym) {
                    sel.value = sym;
                    sel.dispatchEvent(new Event('change'));
                    window.showView && window.showView('overview');
                }
            };
        });
    }
    document.querySelectorAll('#scanView th[data-sort]').forEach((th) => {
        th.classList.toggle('sorted', th.dataset.sort === table.sort);
    });
    const sub = document.getElementById('scanSub');
    if (sub) sub.textContent = `${table.count} instrument(s) · sorted by ${table.sort} · window ${table.window_s}s · ${new Date(table.as_of).toLocaleTimeString()}`;
    const note = document.getElementById('scanNote');
    if (note) note.textContent = table.score_note || '';
}

async function scanRefresh() {
    scanEnsureView();
    if (!document.getElementById('scanRows')) return;
    const inView = (document.querySelector('.view[data-view="scanner"]') || {}).classList?.contains('active');
    if (!inView && SCAN.lastKey) return;                 // only poll the visible view
    try {
        const table = await api(`/api/atlas/scanner?sort=${encodeURIComponent(SCAN.sort)}&limit=60`);
        const key = `${table.as_of && Math.round(table.as_of / 1000)}|${table.sort}|${table.count}|${(table.rows || []).length}`;
        if (key === SCAN.lastKey) return;
        SCAN.lastKey = key;
        scanRender(table);
    } catch (e) {
        const sub = document.getElementById('scanSub');
        if (sub) sub.textContent = `scanner unavailable: ${e}`;
    }
}

(function scanLoop() {
    scanEnsureView();
    SCAN.timer = setInterval(() => {
        if (window.OFAPINTENT && OFAPINTENT.anyHeld()) return;   // never repaint over the user's hands
        const view = document.querySelector('.view[data-view="scanner"]');
        if (view && (view.classList.contains('active') || view.style.display === 'flex')) scanRefresh();
    }, 4000);
    setTimeout(scanRefresh, 5000);
})();

File: orderflow_system/atlas/scanner.py
"""Market scanner — one ranked table across every streaming instrument.

the reference platform's *Market Analyzer* watches a list of instruments and sorts them by a
column. This is the same idea over the order-flow features this program already
computes: it reads each instrument's live analysers and produces one row per symbol,
so "which of my instruments is doing something right now?" is one glance instead of
nine view switches.

Nothing here is a signal: the columns are what the tape, the book and the intent
reader already say, ranked. The composite `score` is an explicit heuristic — it is
documented in the UI tooltip and in the docs, because a ranking that pretends to be
neutral is worse than one that states its weights.
"""

from __future__ import annotations

import time
from typing import Any, Optional

from orderflow_system.atlas import hub as hub_mod


class MarketScanner:
    """Reads a FeatureHub and ranks its instruments."""

    def __init__(self, hub: "hub_mod.FeatureHub", window_s: float = 900.0) -> None:
        self.hub = hub
        self.window_ms = int(max(60.0, float(window_s)) * 1000)
        self._cache: Optional[dict[str, Any]] = None
        self._cache_key: tuple = ()
        self._cache_ms = 0
        self.builds = 0
        self.cache_hits = 0

    # ── per-instrument ────────────────────────────────────────────
    def _row(self, symbol: str, feats: "hub_mod.SymbolFeatures", now_ms: int) -> dict[str, Any]:
        tape = feats.tape
        stats = tape.stats()
        speed = stats.get("speed") or {}
        row: dict[str, Any] = {
            "symbol": symbol,
            "last": 0.0, "chg_pct": None, "high": None, "low": None,
            "volume": stats.get("volume", 0.0),
            "delta": stats.get("delta", 0.0),
            "delta_pct": None,
            "prints": stats.get("prints", 0),
            "prints_per_s": speed.get("per_s") or speed.get("rate") or 0.0,
            "avg_size": 0.0,
            "big_trades": stats.get("big_trades", 0),
            "sweeps": stats.get("sweeps", 0),
            "stop_runs": stats.get("stop_runs", 0),
            "icebergs": stats.get("icebergs", 0),
            "liquidations": stats.get("liquidations", 0),
            "imbalances": 0,
            "pressure_buy": None, "pressure_sell": None, "pressure_side": None, "pressure_edge": None,
            "absorption_side": None, "absorption_score": None,
            "depth_executions": 0, "depth_refills": 0,
            "vwap": None, "ticks_from_vwap": None, "vwap_side": None,
            "score": 0.0,
        }

        # price window from the tape's recent prints (the hub stays the source of truth
        # for price, so the scanner never needs its own feed)
        recent = tape.recent(240)
        if recent:
            first = recent[0]
            last = recent[-1]
            row["last"] = round(float(last["price"]), 8)
            if first.get("price"):
                row["chg_pct"] = round((float(last["price"]) / float(first["price"]) - 1.0) * 100.0, 4)
            prices = [float(p["price"]) for p in recent if p.get("price")]
            if prices:
                row["high"] = round(max(prices), 8)
                row["low"] = round(min(prices), 8)
            if row["prints"]:
                row["avg_size"] = round(float(row["volume"]) / max(1, int(row["prints"])), 6)
        if row["volume"]:
            row["delta_pct"] = round(100.0 * float(row["delta"]) / float(row["volume"]), 2)

        try:
            imb = feats.imbalance.snapshot()
            row["imbalances"] = len(imb.get("levels") or imb.get("rows") or [])
        except Exception:
            pass

        try:
            p = feats.intent.pressure()
            row["pressure_buy"] = p.get("buy_pct_of_normal")
            row["pressure_sell"] = p.get("sell_pct_of_normal")
            row["pressure_side"] = p.get("side")
            row["pressure_edge"] = p.get("edge_pct")
            a = feats.intent.absorption()
            if a.get("absorbing"):
                row["absorption_side"] = a.get("absorbing")
                row["absorption_score"] = a.get("score")
        except Exception:
            pass

        detector = getattr(feats, "detector", None)
        if detector is not None:
            try:
                dstats = detector.stats()
                row["depth_executions"] = dstats.get("executions", 0)
                row["depth_refills"] = dstats.get("refills", 0)
            except Exception:
                pass

        study = getattr(feats, "vwap", None)
        if study is not None:
            try:
                v = study.snapshot(points=8)
                row["vwap"] = v.get("vwap")
                row["ticks_from_vwap"] = v.get("ticks_from_vwap")
                if v.get("vwap") and row["last"]:
                    row["vwap_side"] = "above" if row["last"] > v["vwap"] else "below"
            except Exception:
                pass

        row["score"] = self._score(row)
        return row

    @staticmethod
    def _score(row: dict[str, Any]) -> float:
        """A stated heuristic, not a verdict.

        Weighted pieces, each capped so no single column can dominate:
          * |delta %|                     → who is winning the tape
          * |pressure edge|               → book imbalance vs its own normal
          * absorption score              → aggression that is not working
          * depth executions + refills    → hidden liquidity being interacted with
          * tape speed                    → whether anything is happening at all
        """
        def cap(value: float, top: float) -> float:
            return min(1.0, abs(value) / top) if top else 0.0

        delta = cap(row.get("delta_pct") or 0.0, 40.0)
        edge = cap(row.get("pressure_edge") or 0.0, 80.0)
        absorb = min(1.0, (row.get("absorption_score") or 0.0) / 100.0)
        depth = cap((row.get("depth_executions") or 0) + 1.5 * (row.get("depth_refills") or 0), 20.0)
        speed = cap(row.get("prints_per_s") or 0.0, 25.0)
        return round(100.0 * (0.28 * delta + 0.22 * edge + 0.18 * absorb + 0.17 * depth + 0.15 * speed), 1)

    # ── table ─────────────────────────────────────────────────────
    def snapshot(self, sort: str = "score", limit: int = 50, max_age_ms: int = 2_000) -> dict[str, Any]:
        now_ms = int(time.time() * 1000)
        key = (len(self.hub.symbols), tuple(sorted(self.hub.symbols)), sort, limit,
               sum(int(getattr(f.tape, "version", 0)) for f in self.hub.symbols.values()))
        if self._cache is not None and key == self._cache_key and now_ms - self._cache_ms <= max_age_ms:
            self.cache_hits += 1
            out = dict(self._cache)
            out["cached"] = True
            return out

        rows = []
        for symbol, feats in sorted(self.hub.symbols.items()):
            try:
                rows.append(self._row(symbol, feats, now_ms))
            except Exception:                      # one broken instrument must not blank the table
                continue

        reverse = sort in ("volume", "delta", "delta_pct", "prints_per_s", "big_trades", "score",
                           "depth_executions", "depth_refills", "absorption_score", "imbalances",
                           "sweeps", "stop_runs", "liquidations", "icebergs", "prints")
        def sort_key(r: dict[str, Any]):
            value = r.get(sort)
            if value is None:
                value = float("-inf") if reverse else float("inf")
            return value
        rows.sort(key=sort_key, reverse=reverse)
        if sort == "symbol":
            rows.sort(key=lambda r: r["symbol"])

        out = {
            "as_of": now_ms,
            "window_s": int(self.window_ms / 1000),
            "sort": sort,
            "count": len(rows),
            "rows": rows[: max(1, limit)],
            "score_note": ("score = 0.28·|Δ%| + 0.22·|pressure edge| + 0.18·absorption "
                           "+ 0.17·depth events + 0.15·tape speed, each capped"),
            "cached": False,
        }
        self.builds += 1
        self._cache = out
        self._cache_key = key
        self._cache_ms = now_ms
        return out

    def stats(self) -> dict[str, Any]:
        return {"builds": self.builds, "cache_hits": self.cache_hits,
                "instruments": len(self.hub.symbols), "window_s": int(self.window_ms / 1000)}
